# Script Contents
### -Import Libraries

### -Import Data

### -Create Data Subset

### -Create an If-Statement Function

### -Clean Dirty Prices

### -Create an If-Statement with the loc() Function

### -Create an If-Statement with For-Loops

### -Task 4.7

# Import Libraries

In [2]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Import Data

In [4]:
# Create a path for my main project folder
path = r"C:\Users\jacks\Instacart Basket Analysis 081424"

In [6]:
# Import the latest version of my combined order and product data
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [8]:
df_ords_prods_merge.shape

(32404859, 14)

# Create a Data Subset

In [11]:
# Create a smaller subset of the ords_prods_merge data to avoid potential memory or processing issues
df = df_ords_prods_merge[:1000000]

In [13]:
#Check shape
df.shape

(1000000, 14)

# Create an If-Statement Function

In [16]:
# Create a function that groups rows based on their price
def price_label(row):

    if row['prices'] <= 5:
        return 'Low-range prodcut'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: return 'Not enough data'

In [18]:
# Use the new function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\4171282743.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [20]:
# Check the values in the new column
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673183
Low-range prodcut    314392
High range            12425
Name: count, dtype: int64

In [24]:
# Check the max value in the 'prices' column
df['prices'].max()

99999.0

In [26]:
# Describing the price column since there is an unusually high price and I want to see some info about the price column
df['prices'].describe()

count    1000000.000000
mean          11.396794
std          425.994349
min            1.000000
25%            4.200000
50%            7.400000
75%           11.300000
max        99999.000000
Name: prices, dtype: float64

In [28]:
# I want to see if there are a lot of columsn with the price as 99999 and what product is priced 99999
df[(df['prices'] == 99999)]

order_id  user_id  order_number  orders_day_of_week  \
128911    183964      873             3                   0   
128924   1851256      873             4                   6   
284218   1915696     1893             1                   5   
284229   2763293     1893             2                   4   
284252   2564805     1893             4                   1   
499076    949012     3339             1                   1   
499119    420057     3339             2                   0   
499141   1511997     3339             3                   1   
499158   1153448     3339             4                   4   
499180   2669259     3339             5                   4   
527458    965345     3531            27                   2   
767471   2291230     5034             6                   1   
767485   2324832     5034             7                   6   
953246    323856     6175            37                   2   
960509    842052     6222            10                   1   

        order_hour_of_day  days_since_prior_order  product_id  \
128911                 10                     7.0       33664   
128924                 12                    13.0       33664   
284218                 17                     7.0       33664   
284229                 16                    13.0       33664   
284252                 17                    30.0       33664   
499076                 14                     7.0       33664   
499119                 11                    13.0       33664   
499141                 18                     8.0       33664   
499158                 16                    10.0       33664   
499180                 14                    21.0       33664   
527458                 19                     8.0       33664   
767471                 17                    30.0       33664   
767485                 11                    30.0       33664   
953246                  9                     6.0       33664   
960509                 18                    17.0       33664   

        add_to_cart_order  reordered           product_name  aisle_id  \
128911                 11          0  2 % Reduced Fat  Milk        84   
128924                  8          1  2 % Reduced Fat  Milk        84   
284218                 10          0  2 % Reduced Fat  Milk        84   
284229                  6          1  2 % Reduced Fat  Milk        84   
284252                  3          1  2 % Reduced Fat  Milk        84   
499076                  8          0  2 % Reduced Fat  Milk        84   
499119                 29          1  2 % Reduced Fat  Milk        84   
499141                 14          1  2 % Reduced Fat  Milk        84   
499158                  1          1  2 % Reduced Fat  Milk        84   
499180                  6          1  2 % Reduced Fat  Milk        84   
527458                  2          0  2 % Reduced Fat  Milk        84   
767471                  3          0  2 % Reduced Fat  Milk        84   
767485                  9          1  2 % Reduced Fat  Milk        84   
953246                  6          0  2 % Reduced Fat  Milk        84   
960509                  2          0  2 % Reduced Fat  Milk        84   

        department_id   prices exists price_range  
128911             16  99999.0   both  High range  
128924             16  99999.0   both  High range  
284218             16  99999.0   both  High range  
284229             16  99999.0   both  High range  
284252             16  99999.0   both  High range  
499076             16  99999.0   both  High range  
499119             16  99999.0   both  High range  
499141             16  99999.0   both  High range  
499158             16  99999.0   both  High range  
499180             16  99999.0   both  High range  
527458             16  99999.0   both  High range  
767471             16  99999.0   both  High range  
767485             16  99999.0   both  High range  
953246             16  99999.0   both  High range  
960509   

### There aren't a very large number of rows with the price 99999. The product is the same in every row; 2% Reduced Fat Milk. I am going to describe the product_name column when 2% Reduced Fat milk is present to see if the price is always 99999.

In [31]:
# Describe 2% Reduced Fat milk
df[(df['product_name'] == '2% Reduced Fat Milk')].describe()

order_id      user_id  order_number  orders_day_of_week  \
count  9.650000e+02   965.000000    965.000000          965.000000   
mean   1.650653e+06  3175.236269     19.613472            2.954404   
std    9.907713e+05  1878.043905     21.654176            1.972843   
min    1.154000e+03    38.000000      1.000000            0.000000   
25%    8.029270e+05  1669.000000      5.000000            1.000000   
50%    1.610262e+06  3220.000000     11.000000            3.000000   
75%    2.514363e+06  4693.000000     26.000000            5.000000   
max    3.416917e+06  6468.000000     99.000000            6.000000   

       order_hour_of_day  days_since_prior_order  product_id  \
count         965.000000               965.00000       965.0   
mean           13.597927                10.95544     23909.0   
std             4.272123                 8.90796         0.0   
min             0.000000                 0.00000     23909.0   
25%            10.000000                 4.00000     23909.0   
50%            14.000000                 7.00000     23909.0   
75%            17.000000                15.00000     23909.0   
max            23.000000                30.00000     23909.0   

       add_to_cart_order   reordered  aisle_id  department_id        prices  
count         965.000000  965.000000     965.0          965.0  9.650000e+02  
mean            5.929534    0.746114      84.0           16.0  9.200000e+00  
std             5.726948    0.435459       0.0            0.0  6.220473e-14  
min             1.000000    0.000000      84.0           16.0  9.200000e+00  
25%             2.000000    0.000000      84.0           16.0  9.200000e+00  
50%             4.000000    1.000000      84.0           16.0  9.200000e+00  
75%             8.000000    1.000000      84.0           16.0  9.200000e+00  
max            43.000000    1.000000      84.0           16.0  9.200000e+00

### I think I see an issue. The 2% milk with the price of 99999 has a space in between the '2' and '%' in the product_name. The description of 2% milk (with no space) returns a price of 9.2 only (9.2 is the value for both min and max price). Perhaps the product name 2 % Reduced Fat Milk (with the space) is the milk that has an absurdly high price of 99999.

In [42]:
# Describe 2 % Reduced Fat milk (notice the space between '2' and '%')
df[(df['product_name'] == '2 % Reduced Fat Milk	')].describe()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
count       0.0      0.0           0.0                 0.0                0.0   
mean        NaN      NaN           NaN                 NaN                NaN   
std         NaN      NaN           NaN                 NaN                NaN   
min         NaN      NaN           NaN                 NaN                NaN   
25%         NaN      NaN           NaN                 NaN                NaN   
50%         NaN      NaN           NaN                 NaN                NaN   
75%         NaN      NaN           NaN                 NaN                NaN   
max         NaN      NaN           NaN                 NaN                NaN   

       days_since_prior_order  product_id  add_to_cart_order  reordered  \
count                     0.0         0.0                0.0        0.0   
mean                      NaN         NaN                NaN        NaN   
std                       NaN         NaN                NaN        NaN   
min                       NaN         NaN                NaN        NaN   
25%                       NaN         NaN                NaN        NaN   
50%                       NaN         NaN                NaN        NaN   
75%                       NaN         NaN                NaN        NaN   
max                       NaN         NaN                NaN        NaN   

       aisle_id  department_id  prices  
count       0.0            0.0     0.0  
mean        NaN            NaN     NaN  
std         NaN            NaN     NaN  
min         NaN            NaN     NaN  
25%         NaN            NaN     NaN  
50%         NaN            NaN     NaN  
75%         NaN            NaN     NaN  
max         NaN            NaN     NaN

In [44]:
# Trying to describe 2 % milk (with a space) is not returing anything. I am going to describe 2% milk (without a space)
df[(df['product_name'] == '2% Reduced Fat Milk')]

order_id  user_id  order_number  orders_day_of_week  \
4754     1244041       38             5                   0   
4766      718156       38             6                   6   
4784      894015       38             7                   6   
5274     2315446       42            11                   3   
5319      606677       43             1                   1   
...          ...      ...           ...                 ...   
981983   3325537     6367            11                   6   
981995   3050463     6367            14                   1   
982076   2616150     6367            19                   0   
991199      1154     6418             1                   3   
999232   2830785     6468             4                   0   

        order_hour_of_day  days_since_prior_order  product_id  \
4754                   14                    22.0       23909   
4766                    9                    13.0       23909   
4784                   10                    28.0       23909   
5274                   18                    19.0       23909   
5319                   18                     7.0       23909   
...                   ...                     ...         ...   
981983                 13                    15.0       23909   
981995                 18                    15.0       23909   
982076                 10                     8.0       23909   
991199                 14                     7.0       23909   
999232                 13                    30.0       23909   

        add_to_cart_order  reordered         product_name  aisle_id  \
4754                   14          0  2% Reduced Fat Milk        84   
4766                    7          1  2% Reduced Fat Milk        84   
4784                    8          1  2% Reduced Fat Milk        84   
5274                    2          0  2% Reduced Fat Milk        84   
5319                    4          0  2% Reduced Fat Milk        84   
...                   ...        ...                  ...       ...   
981983                  9          1  2% Reduced Fat Milk        84   
981995                  1          1  2% Reduced Fat Milk        84   
982076                  8          1  2% Reduced Fat Milk        84   
991199                  8          0  2% Reduced Fat Milk        84   
999232                  2          0  2% Reduced Fat Milk        84   

        department_id  prices exists        price_range  
4754               16     9.2   both  Mid-range product  
4766               16     9.2   both  Mid-range product  
4784               16     9.2   both  Mid-range product  
5274               16     9.2   both  Mid-range product  
5319               16     9.2   both  Mid-range product  
...               ...     ...    ...                ...  
981983             16     9.2   both  Mid-range product  
981995             16     9.2   both  Mid-range product  
982076             16     9.2   both  Mid-range product  
991199             16     9.2   both  Mid-range product  
999232             16     9.2   both  Mid-range product  

[965 rows x 15 columns]

### Although there does appear to be a difference in the product_name column values ('2 %' versus '2%'), trying to describe '2 %' is not returning any results. I notice there is also a difference in the product_id column values. '2 %' has a product_id of 33664 while '2%' has a prodcut_id of 23909. Perhaps I can describe the product_id value of 33664 to see if all product_ids with this value have a price of 99999.

In [47]:
#First, I want to view the df with product_id 33664 specified. 
df[(df['product_id'] == 33664)]

order_id  user_id  order_number  orders_day_of_week  \
128911    183964      873             3                   0   
128924   1851256      873             4                   6   
284218   1915696     1893             1                   5   
284229   2763293     1893             2                   4   
284252   2564805     1893             4                   1   
499076    949012     3339             1                   1   
499119    420057     3339             2                   0   
499141   1511997     3339             3                   1   
499158   1153448     3339             4                   4   
499180   2669259     3339             5                   4   
527458    965345     3531            27                   2   
767471   2291230     5034             6                   1   
767485   2324832     5034             7                   6   
953246    323856     6175            37                   2   
960509    842052     6222            10                   1   

        order_hour_of_day  days_since_prior_order  product_id  \
128911                 10                     7.0       33664   
128924                 12                    13.0       33664   
284218                 17                     7.0       33664   
284229                 16                    13.0       33664   
284252                 17                    30.0       33664   
499076                 14                     7.0       33664   
499119                 11                    13.0       33664   
499141                 18                     8.0       33664   
499158                 16                    10.0       33664   
499180                 14                    21.0       33664   
527458                 19                     8.0       33664   
767471                 17                    30.0       33664   
767485                 11                    30.0       33664   
953246                  9                     6.0       33664   
960509                 18                    17.0       33664   

        add_to_cart_order  reordered           product_name  aisle_id  \
128911                 11          0  2 % Reduced Fat  Milk        84   
128924                  8          1  2 % Reduced Fat  Milk        84   
284218                 10          0  2 % Reduced Fat  Milk        84   
284229                  6          1  2 % Reduced Fat  Milk        84   
284252                  3          1  2 % Reduced Fat  Milk        84   
499076                  8          0  2 % Reduced Fat  Milk        84   
499119                 29          1  2 % Reduced Fat  Milk        84   
499141                 14          1  2 % Reduced Fat  Milk        84   
499158                  1          1  2 % Reduced Fat  Milk        84   
499180                  6          1  2 % Reduced Fat  Milk        84   
527458                  2          0  2 % Reduced Fat  Milk        84   
767471                  3          0  2 % Reduced Fat  Milk        84   
767485                  9          1  2 % Reduced Fat  Milk        84   
953246                  6          0  2 % Reduced Fat  Milk        84   
960509                  2          0  2 % Reduced Fat  Milk        84   

        department_id   prices exists price_range  
128911             16  99999.0   both  High range  
128924             16  99999.0   both  High range  
284218             16  99999.0   both  High range  
284229             16  99999.0   both  High range  
284252             16  99999.0   both  High range  
499076             16  99999.0   both  High range  
499119             16  99999.0   both  High range  
499141             16  99999.0   both  High range  
499158             16  99999.0   both  High range  
499180             16  99999.0   both  High range  
527458             16  99999.0   both  High range  
767471             16  99999.0   both  High range  
767485             16  99999.0   both  High range  
953246             16  99999.0   both  High range  
960509   

In [49]:
# All rows with  product_id 33664 are for the 2 % milk priced at 99999. I'm going to describe product_id 33664 to check the price mins/maxs
df[(df['product_id'] == 33664)].describe()

order_id      user_id  order_number  orders_day_of_week  \
count  1.500000e+01    15.000000     15.000000           15.000000   
mean   1.515340e+06  3341.066667      7.733333            2.533333   
std    8.912293e+05  1702.361698     10.305800            2.099887   
min    1.839640e+05   873.000000      1.000000            0.000000   
25%    8.955320e+05  1893.000000      2.500000            1.000000   
50%    1.511997e+06  3339.000000      4.000000            2.000000   
75%    2.308031e+06  4282.500000      6.500000            4.000000   
max    2.763293e+06  6222.000000     37.000000            6.000000   

       order_hour_of_day  days_since_prior_order  product_id  \
count          15.000000               15.000000        15.0   
mean           14.600000               14.666667     33664.0   
std             3.268901                8.949594         0.0   
min             9.000000                6.000000     33664.0   
25%            11.500000                7.500000     33664.0   
50%            16.000000               13.000000     33664.0   
75%            17.000000               19.000000     33664.0   
max            19.000000               30.000000     33664.0   

       add_to_cart_order  reordered  aisle_id  department_id   prices  
count          15.000000  15.000000      15.0           15.0     15.0  
mean            7.866667   0.533333      84.0           16.0  99999.0  
std             6.937133   0.516398       0.0            0.0      0.0  
min             1.000000   0.000000      84.0           16.0  99999.0  
25%             3.000000   0.000000      84.0           16.0  99999.0  
50%             6.000000   1.000000      84.0           16.0  99999.0  
75%             9.500000   1.000000      84.0           16.0  99999.0  
max            29.000000   1.000000      84.0           16.0  99999.0

### At this point, I think it is safe to say that the price of 99999 should be replaced with something more appropriate, such as 9.2, the price of the 2% milk (product_id 23909). I am going to replace price 99999 with price 9.9. I considered replacing price 99999 with  9.2, the price of product_id 23909. However, these items appear to be two different products. The product_ids are different and the product_names are slightly different. I think it makes sense to for the products to have slightly different prices. Perhaps 99999 is a typo that was intended to be 9.9. 

In [54]:
# Use the replace function to replace 99999 with 9.9 in the prices column
df['prices'].replace(99999.0, 9.9, inplace=True)

C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\1067689343.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['prices'].replace(99999.0, 9.9, inplace=True)
C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\1067689343.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['prices'].replace(99999.0, 9.9, inplace=True)


In [56]:
# Describe product_id 33664 to see if the replace worked
df[(df['product_id'] == 33664)].describe()

order_id      user_id  order_number  orders_day_of_week  \
count  1.500000e+01    15.000000     15.000000           15.000000   
mean   1.515340e+06  3341.066667      7.733333            2.533333   
std    8.912293e+05  1702.361698     10.305800            2.099887   
min    1.839640e+05   873.000000      1.000000            0.000000   
25%    8.955320e+05  1893.000000      2.500000            1.000000   
50%    1.511997e+06  3339.000000      4.000000            2.000000   
75%    2.308031e+06  4282.500000      6.500000            4.000000   
max    2.763293e+06  6222.000000     37.000000            6.000000   

       order_hour_of_day  days_since_prior_order  product_id  \
count          15.000000               15.000000        15.0   
mean           14.600000               14.666667     33664.0   
std             3.268901                8.949594         0.0   
min             9.000000                6.000000     33664.0   
25%            11.500000                7.500000     33664.0   
50%            16.000000               13.000000     33664.0   
75%            17.000000               19.000000     33664.0   
max            19.000000               30.000000     33664.0   

       add_to_cart_order  reordered  aisle_id  department_id        prices  
count          15.000000  15.000000      15.0           15.0  1.500000e+01  
mean            7.866667   0.533333      84.0           16.0  9.900000e+00  
std             6.937133   0.516398       0.0            0.0  1.838704e-15  
min             1.000000   0.000000      84.0           16.0  9.900000e+00  
25%             3.000000   0.000000      84.0           16.0  9.900000e+00  
50%             6.000000   1.000000      84.0           16.0  9.900000e+00  
75%             9.500000   1.000000      84.0           16.0  9.900000e+00  
max            29.000000   1.000000      84.0           16.0  9.900000e+00

### Now that the absuredly high price of 99999 has been replaced, I want to recheck the price_range value counts

In [59]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673183
Low-range prodcut    314392
High range            12425
Name: count, dtype: int64

In [61]:
# Perhaps I need to re-apply the price_range function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\4294128843.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [63]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673198
Low-range prodcut    314392
High range            12410
Name: count, dtype: int64

### The High range has reduced by 15, the number of rows that were priced at 99999, so that is good. But there are still more High range counts than I expected. 

In [66]:
# Describing the prices column to see the current max
df['prices'].describe()

count    1000000.000000
mean           9.896958
std          177.499224
min            1.000000
25%            4.200000
50%            7.400000
75%           11.300000
max        14900.000000
Name: prices, dtype: float64

In [68]:
# 14900 is another very high max I did't anticipate. I will investigate by viewing this price
df[(df['prices'] == 14900)]

order_id  user_id  order_number  orders_day_of_week  \
1576      912404       17            12                   2   
1638      603376       17            22                   6   
16522    3264360      135             2                   2   
16528     892534      135             3                   0   
53672     229704      342             8                   1   
...          ...      ...           ...                 ...   
958969    717264     6217            12                   0   
958974    840159     6217            13                   1   
959020   1235750     6217            20                   5   
960274     40712     6221             5                   0   
960348   2567369     6221            10                   2   

        order_hour_of_day  days_since_prior_order  product_id  \
1576                   14                     5.0       21553   
1638                   16                     4.0       21553   
16522                  21                    13.0       21553   
16528                   8                    12.0       21553   
53672                  19                    30.0       21553   
...                   ...                     ...         ...   
958969                 15                     6.0       21553   
958974                 11                    15.0       21553   
959020                  9                    11.0       21553   
960274                 15                     8.0       21553   
960348                 17                    17.0       21553   

        add_to_cart_order  reordered                      product_name  \
1576                    5          0  Lowfat 2% Milkfat Cottage Cheese   
1638                    3          1  Lowfat 2% Milkfat Cottage Cheese   
16522                   6          0  Lowfat 2% Milkfat Cottage Cheese   
16528                   3          1  Lowfat 2% Milkfat Cottage Cheese   
53672                   9          0  Lowfat 2% Milkfat Cottage Cheese   
...                   ...        ...                               ...   
958969                  6          1  Lowfat 2% Milkfat Cottage Cheese   
958974                  4          1  Lowfat 2% Milkfat Cottage Cheese   
959020                  6          1  Lowfat 2% Milkfat Cottage Cheese   
960274                  8          0  Lowfat 2% Milkfat Cottage Cheese   
960348                  3          1  Lowfat 2% Milkfat Cottage Cheese   

        aisle_id  department_id   prices exists price_range  
1576         108             16  14900.0   both  High range  
1638         108             16  14900.0   both  High range  
16522        108             16  14900.0   both  High range  
16528        108             16  14900.0   both  High range  
53672        108             16  14900.0   both  High range  
...          ...            ...      ...    ...         ...  
958969       108             16  14900.0   both  High range  
958974       108             16  14900.0   both  High range  
959020       108             16  14900.0   both  High range  
960274       108             16  14900.0   both  High range  
960348       108             16  14900.0   both  High range  

[142 rows x 15 columns]

In [70]:
# It appears price 14900 is the price for product_id 21553
df[(df['product_id'] == 21553)].describe()

order_id      user_id  order_number  orders_day_of_week  \
count  1.420000e+02   142.000000    142.000000          142.000000   
mean   1.658043e+06  3665.126761     12.042254            2.767606   
std    1.031165e+06  2058.345600     12.297652            1.998813   
min    3.245000e+03    17.000000      1.000000            0.000000   
25%    6.638038e+05  1284.000000      4.000000            1.000000   
50%    1.597053e+06  4415.000000      9.000000            3.000000   
75%    2.685848e+06  5440.000000     14.000000            5.000000   
max    3.411639e+06  6221.000000     65.000000            6.000000   

       order_hour_of_day  days_since_prior_order  product_id  \
count         142.000000              142.000000       142.0   
mean           13.500000               12.570423     21553.0   
std             4.010182                8.822439         0.0   
min             0.000000                0.000000     21553.0   
25%            11.000000                6.250000     21553.0   
50%            14.000000                9.000000     21553.0   
75%            16.000000               17.000000     21553.0   
max            22.000000               30.000000     21553.0   

       add_to_cart_order   reordered  aisle_id  department_id   prices  
count         142.000000  142.000000     142.0          142.0    142.0  
mean            9.507042    0.591549     108.0           16.0  14900.0  
std             7.341465    0.493287       0.0            0.0      0.0  
min             1.000000    0.000000     108.0           16.0  14900.0  
25%             4.000000    0.000000     108.0           16.0  14900.0  
50%             7.000000    1.000000     108.0           16.0  14900.0  
75%            14.000000    1.000000     108.0           16.0  14900.0  
max            38.000000    1.000000     108.0           16.0  14900.0

In [92]:
# There are 142 rows of product_id 21553 and they all have the price of 14900 
# I am going to see if there are any other cottage cheese products 
df[(df['product_name'].str.contains('Cottage'))]

order_id  user_id  order_number  orders_day_of_week  \
825      2115522       10             3                   3   
1576      912404       17            12                   2   
1638      603376       17            22                   6   
6124     1368931       50            45                   1   
6193     3330597       50            53                   3   
...          ...      ...           ...                 ...   
986833   3162681     6402             3                   3   
986884   2950754     6402             5                   2   
992138    476964     6430             2                   2   
998222   2300355     6463            19                   1   
998267   2981876     6463            24                   0   

        order_hour_of_day  days_since_prior_order  product_id  \
825                    19                    12.0       13083   
1576                   14                     5.0       21553   
1638                   16                     4.0       21553   
6124                   12                     6.0       28476   
6193                   11                     2.0       28476   
...                   ...                     ...         ...   
986833                  1                    27.0       13083   
986884                 21                     8.0       28476   
992138                 22                     9.0       13083   
998222                 16                     8.0       13083   
998267                 17                     8.0       13083   

        add_to_cart_order  reordered                       product_name  \
825                    33          0   Lowfat Small Curd Cottage Cheese   
1576                    5          0   Lowfat 2% Milkfat Cottage Cheese   
1638                    3          1   Lowfat 2% Milkfat Cottage Cheese   
6124                    9          0  Organic Small Curd Cottage Cheese   
6193                    4          1  Organic Small Curd Cottage Cheese   
...                   ...        ...                                ...   
986833                 14          0   Lowfat Small Curd Cottage Cheese   
986884                 23          0  Organic Small Curd Cottage Cheese   
992138                 15          0   Lowfat Small Curd Cottage Cheese   
998222                  3          0   Lowfat Small Curd Cottage Cheese   
998267                 10          1   Lowfat Small Curd Cottage Cheese   

        aisle_id  department_id   prices exists        price_range  
825          108             16      9.4   both  Mid-range product  
1576         108             16  14900.0   both         High range  
1638         108             16  14900.0   both         High range  
6124         108             16      8.2   both  Mid-range product  
6193         108             16      8.2   both  Mid-range product  
...          ...            ...      ...    ...                ...  
986833       108             16      9.4   both  Mid-range product  
986884       108             16      8.2   both  Mid-range product  
992138       108             16      9.4   both  Mid-range product  
998222       108             16      9.4   both  Mid-range product  
998267       108             16      9.4   both  Mid-range product  

[2313 rows x 15 columns]

### There are a variety of cottage cheese prodcts with varying prices. This tells me that product)id 21553 is a unique product that could have a unique price from other cottage cheeses. I am going to replace price 14900 with 14.9 since 14900 is too high and is possibly a typo for 14.9.

In [95]:
# Use the replace function to replace 14900 with 14.9 in the prices column
df['prices'].replace(14900.0, 14.9, inplace=True)

C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\2016962084.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['prices'].replace(14900.0, 14.9, inplace=True)
C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\2016962084.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['prices'].replace(14900.0, 14.9, inplace=True)


In [97]:
# Describe product_id 21553 to see if the replace worked
df[(df['product_id'] == 21553)].describe()

order_id      user_id  order_number  orders_day_of_week  \
count  1.420000e+02   142.000000    142.000000          142.000000   
mean   1.658043e+06  3665.126761     12.042254            2.767606   
std    1.031165e+06  2058.345600     12.297652            1.998813   
min    3.245000e+03    17.000000      1.000000            0.000000   
25%    6.638038e+05  1284.000000      4.000000            1.000000   
50%    1.597053e+06  4415.000000      9.000000            3.000000   
75%    2.685848e+06  5440.000000     14.000000            5.000000   
max    3.411639e+06  6221.000000     65.000000            6.000000   

       order_hour_of_day  days_since_prior_order  product_id  \
count         142.000000              142.000000       142.0   
mean           13.500000               12.570423     21553.0   
std             4.010182                8.822439         0.0   
min             0.000000                0.000000     21553.0   
25%            11.000000                6.250000     21553.0   
50%            14.000000                9.000000     21553.0   
75%            16.000000               17.000000     21553.0   
max            22.000000               30.000000     21553.0   

       add_to_cart_order   reordered  aisle_id  department_id        prices  
count         142.000000  142.000000     142.0          142.0  1.420000e+02  
mean            9.507042    0.591549     108.0           16.0  1.490000e+01  
std             7.341465    0.493287       0.0            0.0  3.921819e-14  
min             1.000000    0.000000     108.0           16.0  1.490000e+01  
25%             4.000000    0.000000     108.0           16.0  1.490000e+01  
50%             7.000000    1.000000     108.0           16.0  1.490000e+01  
75%            14.000000    1.000000     108.0           16.0  1.490000e+01  
max            38.000000    1.000000     108.0           16.0  1.490000e+01

In [99]:
# re-apply the price_range function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\98108772.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [101]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673340
Low-range prodcut    314392
High range            12268
Name: count, dtype: int64

In [103]:
# Describing the prices column to see the current max
df['prices'].describe()

count    1000000.000000
mean           7.783274
std            4.233772
min            1.000000
25%            4.200000
50%            7.400000
75%           11.300000
max           25.000000
Name: prices, dtype: float64

## A max price of 25 is reasonable. According to the exercise material, I should have no items in the High range within my subset of 1,000,000 rows. I am not sure why I still have 12268 rows in the High range. I am going to proceed with the exercise with my current dataframe. 

# If-Statements with the loc() Function

In [107]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [109]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [111]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [113]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     673340
Low-range product     314392
High-range product     12268
Name: count, dtype: int64

### Before applying the if-statement with the loc() function to the ords_prods_merge dataframe, I am going to replace the prices 99999 and 14900 with 9.9 and 14.9 since this is what was done in the subset df

In [116]:
df_ords_prods_merge['prices'].replace(99999.0, 9.9, inplace=True)
df_ords_prods_merge['prices'].replace(14900.0, 14.9, inplace=True)

C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\722935586.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords_prods_merge['prices'].replace(99999.0, 9.9, inplace=True)
C:\Users\jacks\AppData\Local\Temp\ipykernel_6712\722935586.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a c

In [118]:
# Check the max price of df_ords_prods_merge to see if there are unusually high prices
df_ords_prods_merge['prices'].max()

25.0

In [120]:
# Apply loc() function if-statement to df_ords_prods_merge
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [122]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [124]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [126]:
# Count values in new column of df_ords_prods_merge
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

# If-Statements with For-Loops

In [129]:
# Find out which day most orders take place
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [133]:
# Create a for-loop that classifies the orders_day_of_week column into different categories
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [135]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [137]:
# Create a new column in df_ords_prods_merge where the for-loop results will be placed
df_ords_prods_merge['busiest_day'] = result

In [139]:
# Check the value counts of the busiest_day column
df_ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 4.7 Task

### 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [143]:
# Find out which days are busiests and least busy
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [210]:
# Create a for-loop that classifies the orders_day_of_week column into the necessary categories
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Two Busiest days")
  elif value == 4 or value == 3:
    result.append("Two slowest days")
  else:
    result.append("Regularly busy")

In [212]:
# Create a new column in df_ords_prods_merge where the for-loop results will be placed
df_ords_prods_merge['busiest_days'] = result

In [214]:
# Check the value counts of the busiest_days column
df_ords_prods_merge['busiest_days'].value_counts(dropna=False)

busiest_days
Regularly busy      12916111
Two Busiest days    11864412
Two slowest days     7624336
Name: count, dtype: int64

#### The value for the two busiest days match the sum of the count of 0 and 1 in the orders_day_of_week column. The same is true for the value of the two slowest days. So, the for-loop appears to have accurately classified the two busiest days, the two slowest days, and all other days into different categories in the busiest_days column.

## 4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [216]:
# First, I need a value count of each hour of the day
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [230]:
# Create a for-loop that classifies the order hours into 3 different categories based on their number of orders
result_hours = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10,11,14,15,13,12,16,9,17]:
    result_hours.append("Most orders")
  elif value in [1,5,2,4,3]:
    result_hours.append("Fewest orders")
  else:
    result_hours.append("Average orders")

##### # I'm grouping the hours based on obvious gaps in the number of orders, such as the gap between 115700 orders at 1 and 218769 at 0.

In [232]:
# Create the new column for this loop
df_ords_prods_merge['busiest_period_of_day'] = result_hours

## 5. Print the frequency for this new column.

In [237]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most orders       23205725
Average orders     8821575
Fewest orders       377559
Name: count, dtype: int64

## 7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [249]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_derived.pkl'))